core/src/main/java/org/elasticsearch/index/query/RandomSampleQuery.java

```diff
@@ -0,0 +1,318 @@
+                            max = Math.min(max, maxDoc);
+                            FakeScorer scorer = new FakeScorer();
+                            scorer.score = score;
+                            collector.setScorer(scorer);
+
+                            int current = min;
+                            while (current < max) {
+                                int gap = getGap(rng);
+                                current = current + gap;
+                                if (current >= maxDoc) {
+                                    return DocIdSetIterator.NO_MORE_DOCS;
+                                }
+                                scorer.doc = current;
+                                if (acceptDocs == null || acceptDocs.get(current)) {
+                                    collector.collect(current);
+                                }
+                            }
+                            return max == maxDoc ? DocIdSetIterator.NO_MORE_DOCS : max;
+                        }
+                        @Override
+                        public long cost() {
+                            return maxDoc;
+                        }
+                    };
+                }
+
+            }
+        };
+    }
+
+    @Override
+    public String toString(String field) {
+        return "RandomSample[p=" + this.p + "](*:*)";
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        RandomSampleQuery other = (RandomSampleQuery)o;
+        return Objects.equals(this.p, other.p) &&
+            Objects.equals(this.seed, other.seed);
+    }
+
+    @Override
+    public int hashCode() {
+        return classHash();
+    }
+
+    /**
+     * A DocIDSetIter that samples on each document.  Empirically, this tends to
+     * be more accurate when the segment is small.  It may also be faster
+     * when the probability is high, since many docs are collected and gap approximation
+     * uses more expensive math
+     */
+    private static class RandomSamplingDocIdSetIterator extends DocIdSetIterator {
+        int doc = -1;
+        final int maxDoc;
+        final Random rng;
+        final double p;
+
+        RandomSamplingDocIdSetIterator(int maxDoc, Random rng, double p) {
+            this.maxDoc = maxDoc;
+            this.rng = rng;
+            this.p = p;
+        }
+
+        @Override
+        public int docID() {
+            return doc;
+        }
+
+        @Override
+        public int nextDoc() throws IOException {
+            return advance(doc + 1);
+        }
+
+        @Override
+        public int advance(int target) throws IOException {
+            doc = target;
+            while (doc < maxDoc) {
+                if (rng.nextDouble() <= p) {
+                    return doc;
+                }
+                doc = doc + 1;
+            }
+            return NO_MORE_DOCS;
```

Comment:
@jpountz Not entirely sure I understand.  Is the issue that the same seeded rng may match different documents depending on how many times it was invoked (due to different scorer positions)?  If that's the case... does it matter?  Every document has the same probability `p` to be matched, indepedendent of other documents.  _Technically_ it's influenced by the prior docs due to not being a truly random generator, but the result is essentially the same from a sampling perspective.

@rjernst Are you saying to re-seed the rng on each document (using it's docid or field, etc)?  Wouldn't that be rather expensive?